src/main/java/com/uw/dao/TrainingDaoImpl.java

```diff
@@ -1,31 +1,170 @@
 package com.uw.dao;
 
 import com.uw.model.Training;
-import com.uw.util.Storage;
+import com.uw.model.TrainingType;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Repository;
 
+import java.time.LocalDate;
+import java.util.List;
+
 @Repository
 public class TrainingDaoImpl implements TrainingDao{
 
-    private Storage storage;
+    private SessionFactory sessionFactory;
 
     @Autowired
-    public void setStorage(Storage storage){
-        this.storage = storage;
+    public void setSessionFactory(SessionFactory sessionFactory){
+        this.sessionFactory = sessionFactory;
     }
 
     @Override
     public void create( Training training ) {
+        Session session = sessionFactory.openSession();
+        Transaction transaction = null;
+        try{
+            transaction = session.beginTransaction();
+            session.persist(training);
+            session.flush();
+            transaction.commit();
+        }catch( Exception e ) {
+            transaction.rollback();
+        }finally {
+            session.close();
+        }
+    }
 
-        long id = this.storage.getSize() + 1;
-        training.setId(id);
-        this.storage.save( training.getId(), training );
+    @Override
+    public Training selectProfile(long id) {
+        Session session = sessionFactory.openSession();
+        Training training = null;
+        try {
+            String hql = "FROM Training t WHERE t.id = :id";
+            training = session.createQuery(hql, Training.class)
+                    .setParameter("id", id)
+                    .uniqueResult();
+        } catch (Exception e) {
+            e.printStackTrace();
+        } finally {
+            session.close();
+        }
+        return training;
+    }
 
+    @Override
+    public List<Training> findAll() {
+        List<Training> trainings = null;
+        try(Session session = sessionFactory.openSession()){
+            trainings = session.createQuery("select t from Training t ", Training.class).getResultList();
+        }catch (Exception e){
+            e.printStackTrace();
+        }
+        return trainings;
     }
 
     @Override
-    public Training selectProfile(long id) {
-        return (Training) this.storage.get(id);
+    public List<Training> getTraineeTrainings(String username, LocalDate fromDate, LocalDate toDate, String trainerName, TrainingType trainingType){
+        Session session = sessionFactory.openSession();
+        Transaction transaction = null;
```

Comment:
There are better alternatives to deal with null checks, Optionals are a good approach but leave it as is , would like to see it on future activities